gunrock/app/hits/hits_enactor.cuh

```diff
@@ -9,426 +9,463 @@
-        //TODO: add all the kernelpolicy settings for all archs
+        GUARD_CU(this -> Init_Threads(this,
+            (CUT_THREADROUTINE)&(GunrockThread<EnactorT>)));
+        return retval;
+    }
 
-        printf("Not yet tuned for this architecture\n");
-        return cudaErrorInvalidDeviceFunction;
+    /**
+      * @brief one run of hits, to be called within GunrockThread
+      * @param thread_data Data for the CPU thread
+      * \return cudaError_t error message(s), if any
+      */
+    cudaError_t Run(ThreadSlice &thread_data)
+    {
+        gunrock::app::Iteration_Loop<
+            // <TODO> change to how many {VertexT, ValueT} data need to communicate
+            //       per element in the inter-GPU sub-frontiers
+            0, 1,
+            // </TODO>
+            IterationT>(
+            thread_data, iterations[thread_data.thread_num]);
+        return cudaSuccess;
     }
 
     /**
-     * @brief HITS Enact kernel entry.
-     *
-     * @tparam HITSProblem HITS Problem type. @see HITSProblem
-     *
-     * @param[in] max_iteration Max iteration number for the algorithm
-     *
-     * \return cudaError_t object which indicates the success of all CUDA function calls.
+     * @brief Reset enactor
+...
+     * @param[in] target Target location of data
+     * \return cudaError_t error message(s), if any
      */
-    cudaError_t Enact(
-        SizeT       max_iteration)
+    cudaError_t Reset(
+        // <TODO> problem specific data if necessary, eg
+        VertexT src = 0,
+        // </TODO>
+        util::Location target = util::DEVICE)
     {
-        int min_sm_version = -1;
-        for (int i=0;i<this->num_gpus;i++)
-            if (min_sm_version == -1 || this->cuda_props[i].device_sm_version < min_sm_version)
-                min_sm_version = this->cuda_props[i].device_sm_version;
-
-        if (min_sm_version >= 300) {
-            return EnactHITS<AdvanceKernelPolicy, FilterKernelPolicy>(
-                    max_iteration);
+        typedef typename GraphT::GpT GpT;
+        cudaError_t retval = cudaSuccess;
+        GUARD_CU(BaseEnactor::Reset(target));
+
+        // This frontier initialization came from the "Hello" app
+        // but it will be overwritten in the iteration loop so that
+        // the frontier contains all nodes
+        for (int gpu = 0; gpu < this->num_gpus; gpu++) {
+           if ((this->num_gpus == 1) ||
+                (gpu == this->problem->org_graph->GpT::partition_table[src])) {
+               this -> thread_slices[gpu].init_size = 1;
+               for (int peer_ = 0; peer_ < this -> num_gpus; peer_++) {
+                   auto &frontier = this -> enactor_slices[gpu * this -> num_gpus + peer_].frontier;
+                   frontier.queue_length = (peer_ == 0) ? 1 : 0;
+                   if (peer_ == 0) {
+                       GUARD_CU(frontier.V_Q() -> ForEach(
+                           [src]__host__ __device__ (VertexT &v) {
+                           v = src;
```

Comment:
You can update this from one of the other apps to see how we queue the whole frontier (all nodes) manually.

Reply:
I'll look around, but any apps in particular?